src/item/Crossbow.php

```diff
@@ -0,0 +1,207 @@
+						-$arrowPitch
+					), $player, true);
+					$entity->setOwningEntity($player);
+
+					if($i !== 1 || $player->isCreative(true)){
+						$entity->setPickupMode(ArrowEntity::PICKUP_CREATIVE);
+					}
+
+					$yawRad = deg2rad($arrowYaw);
+					$pitchRad = deg2rad($arrowPitch);
+					$y = -sin($pitchRad);
+					$xz = cos($pitchRad);
+					$x = -$xz * sin($yawRad);
+					$z = $xz * cos($yawRad);
+
+					$directionVector = (new Vector3($x, $y, $z))->normalize();
+					$entity->setMotion($directionVector->multiply(3.15));
+
+					$entities[] = $entity;
+				}
+			}else{
+				$entity = new ArrowEntity(Location::fromObject(
+					$player->getEyePos(),
+					$player->getWorld(),
+					($location->yaw > 180 ? 360 : 0) - $location->yaw,
+					-$location->pitch
+				), $player, true);
+				$entity->setMotion($player->getDirectionVector()->multiply(3.15));
+				$entity->setOwningEntity($player);
+				$entity->setPickupMode($player->hasFiniteResources() ? ArrowEntity::PICKUP_ANY : ArrowEntity::PICKUP_CREATIVE);
+				$entities[] = $entity;
+			}
+			if(count($entities) === 0){
+				throw new AssumptionFailedError("This should never happen");
+			}
+			$ev = new EntityShootCrossbowEvent($player, $this, $entities);
+			$ev->call();
+			if($ev->isCancelled()){
+				foreach($ev->getProjectiles() as $projectile){
+					$projectile->flagForDespawn();
+				}
+				return ItemUseResult::FAIL;
+			}
+			foreach($ev->getProjectiles() as $projectile){
+				if($projectile instanceof Projectile){ //This might have changed by plugins
+					$launchEv = new ProjectileLaunchEvent($projectile);
+					$launchEv->call();
+					if($launchEv->isCancelled()){
+						$projectile->flagForDespawn();
+						return ItemUseResult::FAIL;
+					}
+					$projectile->spawnToAll();
+				}
+			}
+			$player->getWorld()->addSound($player->getPosition(), new CrossbowShootSound());
+			if($player->hasFiniteResources()){
+				$this->applyDamage($this->hasEnchantment(VanillaEnchantments::MULTISHOT()) ? 3 : 1);
+			}
+			$this->chargedItem = null;
+			return ItemUseResult::SUCCESS;
+		}
+		return ItemUseResult::NONE;
+	}
+
+	public function onReleaseUsing(Player $player, array &$returnedItems) : ItemUseResult{
+		$useDuration = $player->getItemUseDuration();
+		if($useDuration < $this->getChargeDuration()){
+			return ItemUseResult::FAIL;
+		}
+		$arrow = VanillaItems::ARROW();
+		$inventory = match(true){
+			$player->getOffHandInventory()->contains($arrow) => $player->getOffHandInventory(),
+			$player->getInventory()->contains($arrow) => $player->getInventory(),
+			default => null
+		};
+		if($player->hasFiniteResources() && $inventory === null){
+			return ItemUseResult::FAIL;
+		}
+		$inventory?->removeItem($arrow);
+		$this->setCharged($arrow);
+		return ItemUseResult::SUCCESS;
+	}
+
+	public function getChargeDuration() : int{
+		$quickChargeLevel = $this->getEnchantmentLevel(VanillaEnchantments::QUICK_CHARGE());
+		return 25 - (5 * $quickChargeLevel);
```

Comment:
This could return 0 if the enchantment level is 5, which would crash in other parts of the code
It could also return a negative result